File: tools/conformance.py
#!/usr/bin/env python3
"""Run spec/conformance_vectors.json (Python parity with p30inspect conformance)."""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
TIER1 = json.loads((ROOT / "spec" / "tier1_alphabet.json").read_text(encoding="utf-8"))
CHAR_TO_VAL = {entry[0]: entry[1] for entry in TIER1["tier1"]}
TOTATIVES = [1, 7, 11, 13, 17, 19, 23, 29]


def residue(v: int) -> int:
    return ((v % 30) + 30) % 30


def is_coprime(v: int) -> bool:
    from math import gcd

    return gcd(v, 30) == 1


def is_tier1(ch: str) -> bool:
    return ch in CHAR_TO_VAL


def position_value(ch: str, index: int) -> int:
    if ch in CHAR_TO_VAL:
        return CHAR_TO_VAL[ch]
    return index * 30 + TOTATIVES[ord(ch) % 8]


def char_bytes(text: str) -> int:
    return 4 + len(text.encode("utf-8"))


def packed_bytes(text: str) -> int:
    return char_bytes(text) + 4


def pack_positions_5bit(text: str, pad: bool) -> bytes:
    n = len(text)
    out = bytearray((n * 5 + 7) // 8)
    for i, ch in enumerate(text):
        value = residue(position_value(ch, i))
        bit_offset = i * 5
        for bit in range(5):
            if (value >> bit) & 1:
                dst = bit_offset + bit
                out[dst // 8] |= 1 << (dst % 8)
    if pad and len(out) == 102:
        out.append(0)
    return bytes(out)


def huffman_payload(text: str) -> int:
    return len(pack_positions_5bit(text, pad=True))


def huffman_total(text: str) -> int:
    return 430 + huffman_payload(text)


def direct_bytes(text: str) -> int:
    return 4 + len(pack_positions_5bit(text, pad=False)) + 10


def hamming_ops(n_bytes: int) -> int:
    words = (n_bytes * 8 + 63) // 64
    return words * 584


def encode_char(text: str) -> bytes:
    return b"P30C" + text.encode("utf-8")


def decode_char(data: bytes) -> str:
    if not data.startswith(b"P30C"):
        raise ValueError("bad CHAR magic")
    return data[4:].decode("utf-8")


def position_checksum(text: str) -> int:
    total = 0
    for i, ch in enumerate(text):
        total = (total + position_value(ch, i)) & 0xFFFFFFFF
    return total


def encode_packed(text: str) -> bytes:
    return encode_char(text) + position_checksum(text).to_bytes(4, "little")


def decode_packed(data: bytes) -> str:
    if len(data) < 8:
        raise ValueError("packed too short")
    cs = int.from_bytes(data[-4:], "little")
    text = decode_char(data[:-4])
    if position_checksum(text) != cs:
        raise ValueError("checksum mismatch")
    return text


def snapshot(text: str) -> dict:
    n = len(text.encode("utf-8"))
    return {
        "char_count": len(text),
        "utf8_bytes": n,
        "library_ops": n * 3,
        "bios_ops": n * 1,
        "hamming_ops": hamming_ops(n),
        "char_bytes": char_bytes(text),
        "packed_bytes": packed_bytes(text),
        "huffman_total": huffman_total(text),
        "huffman_payload": huffman_payload(text),
        "direct_bytes": direct_bytes(text),
    }


def run_vector(v: dict) -> None:
    kind = v["kind"]
    text = v["text"]
    exp = v.get("expect", {})

    if kind == "tier1_mapping":
        ch = text[0] if text else ""
        if len(text) != 1:
            raise AssertionError("tier1_mapping needs one char")
        got = CHAR_TO_VAL.get(ch)
        if got != exp["tier1_value"]:
            raise AssertionError(f"value {got} != {exp['tier1_value']}")
        if residue(got) != exp["residue_mod30"]:
            raise AssertionError("residue mismatch")
        if not is_coprime(got):
            raise AssertionError("not coprime")

    elif kind == "all_positions_coprime":
        if len(text) != exp["char_count"]:
            raise AssertionError("char count")
        for i, ch in enumerate(text):
            if not is_coprime(position_value(ch, i)):
                raise AssertionError(f"not coprime at {i}")

    elif kind == "char_roundtrip":
        dec = decode_char(encode_char(text))
        if dec != text:
            raise AssertionError("char roundtrip")
        for i, ch in enumerate(text):
            if not is_coprime(position_value(ch, i)):
                raise AssertionError("verify failed")

    elif kind == "packed_roundtrip":
        dec = decode_packed(encode_packed(text))
        if dec != text:
            raise AssertionError("packed roundtrip")

    elif kind == "infoton_demo":
        snap = snapshot(text)
        for key in (
            "char_count",
            "utf8_bytes",
            "library_ops",
            "bios_ops",
            "hamming_ops",
            "char_bytes",
            "packed_bytes",
            "huffman_total",
            "huffman_payload",
            "direct_bytes",
        ):
            if snap[key] != exp[key]:
                raise AssertionError(f"{key}: {snap[key]} != {exp[key]}")
        escapes = [c for c in text if not is_tier1(c)]
        if escapes != exp.get("tier1_escapes", []):
            raise AssertionError(f"escapes {escapes}")

    else:
        raise AssertionError(f"unknown kind {kind}")


def main() -> None:
    path = ROOT / "spec" / "conformance_vectors.json"
    if len(sys.argv) > 1:
        path = Path(sys.argv[1])
    data = json.loads(path.read_text(encoding="utf-8"))
    vectors = data["vectors"]
    failed = []
    for v in vectors:
        try:
            run_vector(v)
        except Exception as e:
            failed.append(f"{v['id']} ({v['kind']}): {e}")
    print(f"=== conformance: {path.name} ===")
    print(f"vectors: {len(vectors)}  passed: {len(vectors) - len(failed)}  failed: {len(failed)}")
    for f in failed:
        print(f"  FAIL {f}")
    if len(vectors) < 100:
        print(f"WARN: expected ≥100 vectors, got {len(vectors)}")
    raise SystemExit(1 if failed else 0)


if __name__ == "__main__":
    main()

File: crates/p30_inspect/src/conformance.rs
use p30_core::ops::CalculatorSnapshot;
use p30_core::storage::{decode_char, decode_packed, encode_char, encode_packed, verify_char_payload};
use p30_core::tier1::{is_tier1, residue_mod30, tier1_value, TIER1_CHARS, TIER1_VALUES};
use p30_core::{position_value, verify_position};
use serde::Deserialize;
use std::path::{Path, PathBuf};

#[derive(Debug, Deserialize)]
pub struct VectorFile {
    pub vector_count: usize,
    pub vectors: Vec<Vector>,
}

#[derive(Debug, Deserialize)]
pub struct Vector {
    pub id: String,
    pub kind: String,
    pub text: String,
    #[serde(default)]
    pub expect: serde_json::Value,
}

pub fn spec_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../spec")
}

pub fn load_vectors(path: &Path) -> Result<VectorFile, String> {
    let data = std::fs::read_to_string(path).map_err(|e| e.to_string())?;
    serde_json::from_str(&data).map_err(|e| e.to_string())
}

pub struct ConformanceReport {
    pub passed: usize,
    pub failed: usize,
    pub failures: Vec<String>,
}

impl ConformanceReport {
    pub fn ok(&self) -> bool {
        self.failed == 0
    }
}

pub fn run_vectors(file: &VectorFile) -> ConformanceReport {
    let mut report = ConformanceReport {
        passed: 0,
        failed: 0,
        failures: Vec::new(),
    };
    for v in &file.vectors {
        match run_one(v) {
            Ok(()) => report.passed += 1,
            Err(msg) => {
                report.failed += 1;
                report.failures.push(format!("{} ({}): {}", v.id, v.kind, msg));
            }
        }
    }
    let _ = file.vector_count; // informational
    report
}

fn run_one(v: &Vector) -> Result<(), String> {
    match v.kind.as_str() {
        "tier1_mapping" => check_tier1_mapping(v),
        "all_positions_coprime" => check_all_coprime(v),
        "char_roundtrip" => check_char_roundtrip(v),
        "packed_roundtrip" => check_packed_roundtrip(v),
        "infoton_demo" => check_infoton_demo(v),
        other => Err(format!("unknown kind {other}")),
    }
}

fn check_tier1_mapping(v: &Vector) -> Result<(), String> {
    let ch = v.text.chars().next().ok_or("empty text")?;
    if v.text.chars().count() != 1 {
        return Err("tier1_mapping expects one character".into());
    }
    let exp_val = v.expect["tier1_value"].as_u64().ok_or("missing tier1_value")? as u32;
    let exp_res = v.expect["residue_mod30"].as_u64().ok_or("missing residue_mod30")? as u8;
    let got = tier1_value(ch).ok_or("not tier1")?;
    if got != exp_val {
        return Err(format!("value {got} != {exp_val}"));
    }
    if residue_mod30(got) != exp_res {
        return Err(format!("residue {} != {}", residue_mod30(got), exp_res));
    }
    if !verify_position(got) {
        return Err("not coprime".into());
    }
    Ok(())
}

fn check_all_coprime(v: &Vector) -> Result<(), String> {
    let exp_count = v.expect["char_count"].as_u64().ok_or("missing char_count")? as usize;
    if v.text.chars().count() != exp_count {
        return Err(format!("char count {}", v.text.chars().count()));
    }
    for (i, ch) in v.text.chars().enumerate() {
        if !verify_position(position_value(ch, i)) {
            return Err(format!("not coprime at {i}"));
        }
    }
    Ok(())
}

fn check_char_roundtrip(v: &Vector) -> Result<(), String> {
    let enc = encode_char(&v.text);
    let dec = decode_char(&enc).map_err(|e| e.to_string())?;
    if dec != v.text {
        return Err("decode mismatch".into());
    }
    if !verify_char_payload(&v.text) {
        return Err("verify failed".into());
    }
    Ok(())
}

fn check_packed_roundtrip(v: &Vector) -> Result<(), String> {
    let enc = encode_packed(&v.text);
    let dec = decode_packed(&enc).map_err(|e| e.to_string())?;
    if dec != v.text {
        return Err("decode mismatch".into());
    }
    Ok(())
}

fn check_infoton_demo(v: &Vector) -> Result<(), String> {
    let snap = CalculatorSnapshot::from_text(&v.text);
    let mut check_usize = |field: &str, got: usize| -> Result<(), String> {
        let want = v.expect[field]
            .as_u64()
            .ok_or_else(|| format!("missing {field}"))? as usize;
        if got != want {
            return Err(format!("{field}: {got} != {want}"));
        }
        Ok(())
    };
    check_usize("char_count", snap.char_count)?;
    check_usize("utf8_bytes", snap.utf8_bytes)?;
    check_usize("library_ops", snap.library.total() as usize)?;
    check_usize("bios_ops", snap.bios.total() as usize)?;
    check_usize("hamming_ops", snap.hamming_ops as usize)?;
    check_usize("char_bytes", snap.char_storage_bytes)?;
    check_usize("packed_bytes", snap.packed_storage_bytes)?;
    check_usize("huffman_total", snap.huffman_total_bytes)?;
    check_usize("huffman_payload", snap.huffman_payload_bytes)?;
    check_usize("direct_bytes", snap.direct_storage_bytes)?;
    if let Some(arr) = v.expect.get("tier1_escapes").and_then(|x| x.as_array()) {
        let escapes: Vec<char> = v.text.chars().filter(|&c| !is_tier1(c)).collect();
        let exp: Vec<char> = arr
            .iter()
            .filter_map(|x| x.as_str().and_then(|s| s.chars().next()))
            .collect();
        if escapes != exp {
            return Err(format!("escapes {:?} != {:?}", escapes, exp));
        }
    }
    Ok(())
}

/// Verify embedded Tier-1 table matches spec/tier1_alphabet.json length.
pub fn verify_tier1_table_integrity() -> Result<(), String> {
    if TIER1_CHARS.len() != 68 || TIER1_VALUES.len() != 68 {
        return Err("tier1 table length != 68".into());
    }
    for (&ch, &val) in TIER1_CHARS.iter().zip(TIER1_VALUES.iter()) {
        if tier1_value(ch) != Some(val) {
            return Err(format!("table mismatch for {ch:?}"));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn conformance_vectors_all_pass() {
        verify_tier1_table_integrity().unwrap();
        let path = spec_dir().join("conformance_vectors.json");
        let file = load_vectors(&path).expect("load vectors");
        assert!(file.vector_count >= 100, "need ≥100 vectors");
        let report = run_vectors(&file);
        for f in &report.failures {
            eprintln!("FAIL: {f}");
        }
        assert_eq!(report.failed, 0, "{} failures", report.failed);
        assert!(report.passed >= 100);
    }
}
